### PR TITLE
Various Help and Parsing features and fixes

### PR DESCRIPTION
**Pull request by hansl**
Added a new `usageNotes` field to command descriptions, moved suboptions to subcommands which are tied to options (removing the awkward `suboption` type), removed unused field `type` from the interface, added enums and multiple types support to the argument parser, and moved `--help-json` to `--help=json`.

### CODE REVIEW

File: packages/angular/cli/utilities/json-schema.ts [hunk 1]
```diff
@@ -73,13 +73,29 @@ export async function parseJsonSchemaToCommandDescription(
     const ldPath = resolve(dirname(jsonPath), schema.$longDescription);
     longDescription = readFileSync(ldPath, 'utf-8');
   }
+  let usageNotes = '';
+  if (typeof schema.$usageNotes == 'string' && schema.$usageNotes) {
+    const unPath = resolve(dirname(jsonPath), schema.$usageNotes);
+    usageNotes = readFileSync(unPath, 'utf-8');
```
Comment (clydin) [inline packages/angular/cli/utilities/json-schema.ts]: error handling if the file does not exist?

File: packages/angular/cli/utilities/json-schema.ts [hunk 2]
```diff
@@ -73,13 +73,29 @@ export async function parseJsonSchemaToCommandDescription(
...
   const type = _getEnumFromValue(schema.$type, CommandType, CommandType.Default);
   const description = '' + (schema.description === undefined ? '' : schema.description);
   const hidden = !!schema.$hidden;
 
-  return { name, description, longDescription, hidden, type, options, aliases, scope, impl };
+  return {
+    name,
+    description,
+    ...(longDescription ? { longDescription } : {}),
+    ...(usageNotes ? { usageNotes } : {}),
```
Comment (clydin) [inline packages/angular/cli/utilities/json-schema.ts]: are the spread operators needed here?

File: packages/angular_devkit/core/src/json/schema/utility.ts [hunk 3]
```diff
@@ -23,6 +24,30 @@ export function getTypesOfSchema(schema: JsonObject | true): Set<string> {
     potentials = new Set([schema.type]);
   } else if (Array.isArray(schema.type)) {
     potentials = new Set(schema.type as string[]);
+  } else if (isJsonArray(schema.enum)) {
+    // Gather the type of each enum values, and use that as a starter for potential types.
+    const filteredTypes = [...schema.enum].map(v => {
```
Comment (clydin) [inline packages/angular_devkit/core/src/json/schema/utility.ts]: for/of would probably be cleaner here and reduce the iterations from four to one.

File: packages/angular/cli/commands/definitions.json [hunk 4]
```diff
@@ -30,12 +30,9 @@
...
         "help": {
-          "type": "boolean",
-          "description": "Shows a help message."
-        },
-        "helpJson": {
-          "type": "boolean",
-          "description": "Shows the metadata associated with each flags, in JSON format."
+          "type": ["boolean", "string"],
+          "description": "Shows a help message. Use '--help=json' as a value to output the help in JSON format.",
+          "default": false
```
Comment (clydin) [inline packages/angular/cli/commands/definitions.json]: `"enum": [true, "json"]`?

or more maybe:
```
  "anyOf": [
    {
      "enum": [ true, "json" ]
    },
    {
      "pattern": "^[Jj][Ss][Oo][Nn]$"
    }
  ]
```

Comment (hansl) [reply packages/angular/cli/commands/definitions.json]: The first one would not handle the `—no-help` case. The second one wouldn’t list the enumerations properly.

File: packages/angular/cli/utilities/json-schema.ts (hunk 2 above)
Comment (hansl) [reply packages/angular/cli/utilities/json-schema.ts]: I don’t want the keys to be in the object, even if the value is undefined.

File: packages/angular/cli/commands/definitions.json (hunk 4 above)
Comment (clydin) [reply packages/angular/cli/commands/definitions.json]: the tsconfig schema uses that structure to handle casing and vscode displays the enum values (not sure what its code is doing).

Comment (clydin) [reply packages/angular/cli/commands/definitions.json]: I'm ok with it either way though.

File: packages/angular/cli/utilities/json-schema.ts (hunk 1 above)
Comment (hansl) [reply packages/angular/cli/utilities/json-schema.ts]: Done.

File: packages/angular_devkit/core/src/json/schema/utility.ts (hunk 3 above)
Comment (hansl) [reply packages/angular_devkit/core/src/json/schema/utility.ts]: Done. It does look cleaner.

File: packages/angular/cli/commands/definitions.json (hunk 4 above)
Comment (hansl) [reply packages/angular/cli/commands/definitions.json]: I would think vscode is showing it from the type itself? The `.d.ts` is cleaner.

I also changed the code of the parser to reject values that are not part of the enum (and adding it to leftovers). I wasn't sure what sort of behaviour would be correct here, TBH. If we set it and it's not part of the enum, then I would consider that a breach of contract. Ignoring the flag might be slightly better but lead to user errors.

I could use some whiteboarding on this actually.

**Pull request merged**